### ISSUE TITLE
KV Watch | some keys missing when use KV api to watch a subject pattern due to wrong NumPending from the server

### ISSUE THREAD

**Issue by pricelessrabbit** [nats-io/nats-server#4824]
#### preconditions
- A KV bucket with about 30k different keys. (subjects structure like `root.<object-id>.<attribute>`)
- During the test, no PUT operations in store are performed. Only watch to get existent values.
- A pattern to watch (like `root.*.*` or `root.object_id_1.*`). 

#### Steps to reproduce

1. Using a watch with `nil` guard, get all the keys that match the pattern and exit

```go
w, _ := kv.Watch("root.object_id_1.*", nats.IgnoreDeletes())

// count keys and save the key names on file
i := 1
f, _  := os.Create("./subject_names")

for msg := range w.Updates(){
	fmt.Println(i)
	if msg == nil {
		break
	}
	f.Write([]byte(msg.Key() + "\n"))
	i++
}
```

final keys count (on my data): 5397

2. Run the watch without breaking at nil value but skipping it and continue to process incoming keys

```go
w, _ := kv.Watch("root.object_id_1.*", nats.IgnoreDeletes())

// count keys and save the key names on file
i := 1
f, _  := os.Create("./subject_names")

for msg := range w.Updates(){
	fmt.Println(i)
	if msg == nil {
		// skip nil and continue watch instead of break
		continue 
	}
	f.Write([]byte(msg.Key() + "\n"))
	i++
}

```

final keys count: 5466
(after halting the never-ending for-range)

In the 2 runs only a single watcher is used and no other operations are performed in the store.

---

Trying to debug, i found out that  Watch (and also Keys) methods, relays on NumPending consumer info attribute as the condition to send the `nil` guard ( to know when all the existent keys have been sent to the channel).

In my case, that count (coming from the server) is wrong so a lot of keys are missing from the watch. 
I have this behaviour only in a specific stream, but i cannot reproduce it in general. I tried to create different kv stores and perform random operations to reproduce the bug with no success.

As a side note:
I have different behaviours based on the nats server versions:
- version `>=` 2.10.4 --> count is wrong but skipping the nil guard gives me all the keys
- version `<` 2.10.4 --> count is always wrong, and the set of returned keys is different (so different server version gives me a different set of keys, but never all of them)

If you need it for debug purposes, i can provide you privately the backup of the stream

### Expected behavior

I expect that when i run a watch and break out on `nil` guard, all existent values in the stream are already sent to the client

### Server and client version

the issue originated from a server with `2.10.2 <= version <= 2.10.4`
i have the issue on versions from 10.0.0 to 10.0.5 and client version 1.31

### Host environment

both on:
- linux with server on docker
- wsl

**Comment by derekcollison** [comment]
Could you check on the latest server 2.10.6 and report if you still are seeing the issue?

**Comment by pricelessrabbit** [comment]
tested on 2.10.6:
- imported stream with issue
- run the same steps described in STR

Result:
In this case i always have 5397 entries both breaking or continuing on nil guard. 

So basically in 2.10.6 i have no way to get all the keys in the store / know how many subjects exist in the stream, but if i use KV GET api to for a specific, single key (like `nats kv get <bucket> <watch missing key>`), the correct value is returned

If i ask for the key history, i also got the value (history is always of 1 PUT operation)

**Comment by pricelessrabbit** [comment]
Thanks :)
As a side note, I have the same wrong count also using the `ConsumerInfo` to get the `NumPending` field. If you need the stream backup with the issue, tell me how can i send it to you.

**Comment by derekcollison** [comment]
Email works, feel free to send it to me with some basic instructions on what I should focus on.
derek@synadia.com

**Comment by derekcollison** [comment]
Will take a look. Thanks.

**Comment by Jarema** [comment]
Hey @pricelessrabbit I think I miss the permissions for the repo you linked as I get page not found error.

**Comment by pricelessrabbit** [comment]
@Jarema just added you and piotrpio to the repo

**Comment by Jarema** [comment]
@pricelessrabbit We did some investigation and this is what we found so far:

1. If you restore the stream using server 2.10.5, you will get proper number of messages, but while having a watcher over them, duplicates will be returned. That confuses the client and is a reason why you are getting messages after the `nil`.
2. If you retore stream using 2.10.6 or upgrade the server with existing stream, no duplicates are returned and `nil` is returned when it should be.

We checked that there are 5396 messages on the Stream for the subject from your reproduction, and that number matches the number returned by your `main.go` example.

To sum it up **- the 2.10.6 seems to work correctly** and also fix sending duplicates that confuses clients and you :).

**Comment by piotrpio** [comment]
@pricelessrabbit Those keys are missing when you break on nil on v2.10.5 because previous duplicated entries "took their spot" - try running `sort sub_names | uniq -cd` and you'll see all the duplicated keys. Since the client is checking `NumPending` in `Watch()`, and `NumPending` is indeed correct, duplicates count towards the returned amount and the keys you mentioned are sent only after `nil`.

If you upgrade your nats-server to v2.10.6 and `Watch()` again, you should see the same number of keys (5396) but no duplicates and all keys (e.g. `ADJ_PH_MIS_3`) should be there.

**Comment by pricelessrabbit** [comment]
oh 😄  now i got it! I tested 2.10.6 (5 days ago) but i didn't check that the keys were all present,  because i got confused by the "wrong" count, and assumed that some were missing. Just tested now 2.10.6 and the issue seems solved.

Thank you very much guys 🎉

**Comment by Jarema** [comment]
Happy we could help!
Closing the issue.
